cranelift/assembler-x64/meta/src/instructions.rs

```diff
@@ -44,5 +45,104 @@ pub fn list() -> Vec<Inst> {
     all.extend(sub::list());
     all.extend(xor::list());
     all.extend(unpack::list());
+
+    // Automatically assign AVX alternates to SSE instructions (see
+    // `Inst::alternate`). This allows later code generation to
+    // instruction-select between the AVX and SSE versions of an instruction.
+    assign_avx_alternates(&mut all);
+
     all
 }
+
+/// Assigns AVX alternates to SSE instructions.
```

Comment:
It's neat that this was made to work automagically -- all the same, I worry a tiny bit that it may be brittle (unexpectedly determine a match when some subtle condition means they're not quite equivalent; or on the other side, miss a match because of something). The note below about rounding and the fact this limits things to just one format for now as a result seems to bear this out a little.

In the spirit of the explicit DSL, I wonder if it would be better to have a clause on the instruction builder that sets the alternate? Something like

```
  inst("vpadd", ...).avx_for("padd"),
  inst("padd", ...),
  ...
```

(`avx_for`, `upgrades`, `equiv`, pick a verb phrase...)

Thoughts?

Reply:
Yeah, I did consider adding an explicit version: `inst(...).alt("vpadd")`. I can't remember what dissuaded me, but probably laziness (there will be a lot of places to update!). But it is more clear so I'm in favor of that &mdash; less magic. I'll rework this to use a DSL function and use the current "check the SSE-to-AVX link" logic to verify that the user has specified `.alt(...)` correctly.